specifyweb/frontend/js_src/lib/components/Statistics/index.tsx

```diff
@@ -0,0 +1,332 @@
+import { removeItem, replaceItem } from '../../utils/utils';
+import { H2 } from '../Atoms';
+import { Button } from '../Atoms/Button';
+import { className } from '../Atoms/className';
+import { useQueries } from '../Toolbar/Query';
+import { usePref } from '../UserPreferences/usePref';
+import { userInformation } from '../InitialContext/userInformation';
+import { awaitPrefsSynced } from '../UserPreferences/helpers';
+import { softFail } from '../Errors/Crash';
+import { Form } from '../Atoms/Form';
+import { Submit } from '../Atoms/Submit';
+import { useDefaultLayout, useDefaultStatsToAdd, useStatsSpec } from './hooks';
+import { Categories } from './Categories';
+import { AddStatDialog } from './AddStatDialog';
+import { PageNameDialog } from './PageNameDialog';
+import { PageButton } from './Buttons';
+
+export function StatsPage(): JSX.Element {
+  const [customLayout, setLayout] = usePref(
+    'statistics',
+    'appearance',
+    'layout'
+  );
+  const statsSpec = useStatsSpec();
+  const defaultLayout = useDefaultLayout(statsSpec);
+  const layout = customLayout ?? defaultLayout;
+  const [state, setState] = React.useState<
+    | State<'EditingState'>
+    | State<
+        'AddingState',
+        {
+          readonly pageIndex: number;
+          readonly categoryIndex: number;
+        }
+      >
+    | State<'DefaultState'>
+    | State<
+        'PageRenameState',
+        {
+          readonly pageIndex: number | undefined;
+        }
+      >
+  >({ type: 'DefaultState' });
+
+  const isAddingItem = state.type === 'AddingState';
+  const isEditing =
+    state.type === 'EditingState' ||
+    isAddingItem ||
+    state.type === 'PageRenameState';
+  const [activePageIndex, setActivePageIndex] = React.useState<number>(0);
+  const defaultStatsToAdd = useDefaultStatsToAdd(
+    layout[activePageIndex],
+    defaultLayout
+  );
+  const filters = React.useMemo(
+    () => ({
+      specifyUser: userInformation.id,
+    }),
+    []
+  );
+  const queries = useQueries(filters, false);
+  const previousLayout = React.useRef(layout);
+
+  return (
+    <Form
+      className={className.containerFullGray}
+      onSubmit={(): void => {
+        setState({ type: 'DefaultState' });
+        awaitPrefsSynced().catch(softFail);
+      }}
+    >
+      <div className="flex items-center gap-2">
+        <H2 className="text-2xl">{commonText('statistics')}</H2>
+        <span className="-ml-2 flex-1" />
+        {isEditing ? (
+          <>
+            <Button.Red onClick={(): void => setLayout(defaultLayout)}>
+              {commonText('reset')}
+            </Button.Red>
+            <Button.Red
+              onClick={(): void => {
+                setLayout(previousLayout.current);
+                setState({ type: 'DefaultState' });
+                setActivePageIndex(
+                  activePageIndex >= previousLayout.current.length
+                    ? previousLayout.current.length - 1
+                    : activePageIndex
```

Comment:
A better way to express this is `Math.min(activePageIndex, previousLayout.current.length - 1)`

It would also be less buggy, as it is easy to confuse `<=`, `>=`, `>`, `<` (there have been a lot of specify bugs caused by that)